### PR TITLE
gh-106290: Fix edge cases around uops

### PR DESCRIPTION
**Pull request by gvanrossum**
A bit of an omnibus PR, alas (let me know if you want it split up). This does roughly the following (though not in that order):

- Tweak uops debugging output
- Fix the bug from gh-106290
- Rename `SET_IP` to `SAVE_IP` (per https://github.com/faster-cpython/ideas/issues/558)
- Add a `SAVE_IP` uop at the start of the trace (ditto)
- Allow `unbound_local_error`; this gives us uops for `LOAD_FAST_CHECK`, `LOAD_CLOSURE`, and `DELETE_FAST`
- Longer traces
- Support `STORE_FAST_LOAD_FAST`, `STORE_FAST_STORE_FAST`
- Add deps on pycore_uops.h to Makefile(.pre.in)

### CODE REVIEW

**Comment by gvanrossum** [comment]
While we're cleaning up uops, maybe we should change "trace" to "superblock"? There are already several other things using the name "trace", e.g. `sys.settrace()`, `tracemalloc`, `lltrace`.

**Comment by gvanrossum** [comment]
This (with uops on by default) benchmarks [2% slower than main](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20230701-3.13.0a0-5b85a49/bm-20230701-linux-x86_64-gvanrossum-tweak_uops-3.13.0a0-5b85a49-vs-base.png) (with uops off by default).

I will now revert the final (temporary) commit that turns on uops by default and mark it as ready for review.

**Comment by gvanrossum** [comment]
(Also, the tests that fail with uops on by default are largely expected. There's one [refleak](https://github.com/python/cpython/actions/runs/5433235300/jobs/9880794480) -- maybe I should ask Kumar to [review](https://github.com/python/cpython/pull/105924#pullrequestreview-1499483051) it again. :-)

**Comment by gvanrossum** [comment]
Maybe the leak is the optimizer created and installed initially.

File: Python/optimizer.c [hunk 1]
```diff
@@ -388,49 +404,45 @@ translate_bytecode_to_trace(
...
+                PyUnicode_AsUTF8(code->co_qualname),
+                PyUnicode_AsUTF8(code->co_filename),
+                code->co_firstlineno,
+                (long)(instr - (_Py_CODEUNIT *)code->co_code_adaptive));
     }
-    return trace_length;
+    return 0;
 
 #undef ADD_TO_TRACE
+#undef ADD_TO_TRACE_BASIC
```
Comment (brandtbucher) [inline Python/optimizer.c]: I don't see where this is defined...

**Review by brandtbucher — APPROVED** [review]
The code itself looks good, haven't tried running it myself or anything. Just one note:

File: Python/optimizer.c (hunk 1 above)
Comment (gvanrossum) [reply Python/optimizer.c]: Turd from the last cleanup. I'll get rid of it, then land it.

**Pull request merged**